### ISSUE TITLE
Non-robust SCG tests

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1499]
The following tests tend to fail when using alternative external BLAS/LAPACK implementations:

```
	251 - igraph_scg_grouping3 (Failed)
	254 - igraph_scg_semiprojectors2 (Failed)
```

Tested with OpenBLAS on macOS and MKL on Linux. It's always these two tests. They seem not to be robust, there is probably some degeneracy.

<details><summary>MKL on Linux:</summary>

```
> cat test_igraph_scg_grouping3.diff
--- /home/szhorvat/Repos/igraph/examples/simple/igraph_scg_grouping3.out	2020-10-24 15:03:18.161052000 +0200
+++ /home/szhorvat/Repos/igraph/build/tests/test_igraph_scg_grouping3.out	2020-10-24 15:59:33.453802000 +0200
@@ -1,5 +1,5 @@
 0.433013 0.57735 0.57735 0.144338 0.144338 0.144338 0.144338 0.144338 0.144338 0.144338
-0 1 1 0 2 2 2 2 2 2
+0 1 1 0 1 2 2 2 2 2
 0 1 1 2 1 1 1 1 1 1
 0 1 1 2 1 1 1 1 1 1
 0 1 1 2 3 3 3 3 3 3
```

```
> cat test_igraph_scg_semiprojectors2.diff
--- /home/szhorvat/Repos/igraph/examples/simple/igraph_scg_semiprojectors2.out	2020-10-24 15:03:18.176041000 +0200
+++ /home/szhorvat/Repos/igraph/build/tests/test_igraph_scg_semiprojectors2.out	2020-10-24 15:59:33.645804000 +0200
@@ -1,11 +1,11 @@
 ----------------------
 0.75 0 0 0.25 0 0 0 0 0 0
-0 0.5 0.5 0 0 0 0 0 0 0
-0 0 0 0 0.166667 0.166667 0.166667 0.166667 0.166667 0.166667
+0 0.444444 0.444444 0 0.111111 0 0 0 0 0
+0 0 0 0 0 0.2 0.2 0.2 0.2 0.2
 ---
 1 0 0 1 0 0 0 0 0 0
-0 1 1 0 0 0 0 0 0 0
-0 0 0 0 1 1 1 1 1 1
+0 1 1 0 1 0 0 0 0 0
+0 0 0 0 0 1 1 1 1 1
 ---
 ----------------------
 1 0 0 0 0 0 0 0 0 0
```
</details>

<details><summary>OpenBLAS on macOS:</summary>

```
$ cat test_igraph_scg_grouping3.diff
--- /Users/szhorvat/Repos/igraph-main/igraph/examples/simple/igraph_scg_grouping3.out	2020-05-26 09:56:54.000000000 +0200
+++ /Users/szhorvat/Repos/igraph-main/igraph/build/tests/test_igraph_scg_grouping3.out	2020-10-24 18:16:45.000000000 +0200
@@ -1,5 +1,5 @@
 0.433013 0.57735 0.57735 0.144338 0.144338 0.144338 0.144338 0.144338 0.144338 0.144338
-0 1 1 0 2 2 2 2 2 2
+0 1 1 0 2 1 2 2 2 2
 0 1 1 2 1 1 1 1 1 1
 0 1 1 2 1 1 1 1 1 1
 0 1 1 2 3 3 3 3 3 3
```

```
$ cat test_igraph_scg_semiprojectors2.diff 
--- /Users/szhorvat/Repos/igraph-main/igraph/examples/simple/igraph_scg_semiprojectors2.out	2020-05-26 09:56:54.000000000 +0200
+++ /Users/szhorvat/Repos/igraph-main/igraph/build/tests/test_igraph_scg_semiprojectors2.out	2020-10-24 18:16:45.000000000 +0200
@@ -1,11 +1,11 @@
 ----------------------
 0.75 0 0 0.25 0 0 0 0 0 0
-0 0.5 0.5 0 0 0 0 0 0 0
-0 0 0 0 0.166667 0.166667 0.166667 0.166667 0.166667 0.166667
+0 0.444444 0.444444 0 0 0.111111 0 0 0 0
+0 0 0 0 0.2 0 0.2 0.2 0.2 0.2
 ---
 1 0 0 1 0 0 0 0 0 0
-0 1 1 0 0 0 0 0 0 0
-0 0 0 0 1 1 1 1 1 1
+0 1 1 0 0 1 0 0 0 0
+0 0 0 0 1 0 1 1 1 1
 ---
 ----------------------
 1 0 0 0 0 0 0 0 0 0
```
</details>

**Comment by vtraag** [comment]
WIth OpenBLAS on Linux the errors are identical to MKL on Linux that you report @szhorvat. However, with MKL, I don't get these errors. For this setup I use `-DBLA_VENDOR=Intel10_64lp` with `libmkl-dev:amd64 (2020.0.166-1)` on Ubuntu 20.04, building with only `BLAS` as external dependency.

**Comment by vtraag** [comment]
There is `cs_realloc`, which seems to work well, I'm double checking this now.

**Comment by szhorvat** [comment]
As expected, the same two tests fail with MSYS2/MinGW as well if using OpenBLAS.

**Comment by vtraag** [comment]
I can confirm the failures on MSYS2/MinGW for both shared and static builds.

**Comment by szhorvat** [comment]
The same also fails when building the internal libraries with x87 FPU instructions instead of SSE:

https://travis-ci.org/github/igraph/igraph/jobs/741864231#L1370

**Comment by szhorvat** [comment]
@gaborcsardi Are you still in contact with the original author of the SCG stuff? The SCG tests have been a pain as they tend to be very sensitive to numerical errors. It is unclear to me what sort of input is needed so that the output would be uniquely defined. It would save a lot of time for us (i.e. reading and thinking through the papers) if the original author could give a hint about this.

`igraph_scg_grouping3` had a very symmetric tree, so naively I tried to replace it with a tree that has no non-trivial automorphisms (and no degenerate eigenvalues), but no luck.

Generally, the issue is that the results depend on the floating point instruction set used (x87 vs SSE) as well as on the specific LAPACK implementation (NetLib, Apple, OpenBLAS, MKL, etc.)

**Comment by szhorvat** [comment]
There were two separate issues causing a test failure. 

The first one, which only occurred when using x87 math, was that the `p` vector was calculated as the leading eigenvector of the stochastic matrix. The corresponding eigenvalue is 1 and the vector's elements all have the same sign. However, some LAPACKs on some platforms may return all-negative values instead of all-positive.

This is now addressed: if the values are negative, we negate the `p` vector. Additionally, `igraph_scg_grouping()` now checks that all `p` values are non-negative. According to the documentation, `p` is a vector of probabilities, thus this makes sense. However, it is unclear to me if the implementation actually requires this.

I note that when the graph is undirected, `p` is simply proportional to the degrees. Thus an easier way to compute it would be 

```
igraph_degree(&g, &p, igraph_vss_all(), IGRAPH_ALL, IGRAPH_LOOPS);
```

This could also be used in `if (do_p)` part of the `igraph_scg_stochastic()` function, only when the graph is undirected. This change is not made at the moment.

**Comment by szhorvat** [comment]
The second problem is different: the partitions that are returned seem to depend on the LAPACK that we use.  Specifically, in the `igraph_scg_grouping3` test, this concerns the `V` 1-by-n matrix which contains the 3rd eigenvector of the stochastic matrix. 

This is what various LAPACK implementations return:

```
Vendored LAPACK (NetLib):
-0.38851434494290565
0.09712858623572650
0.09712858623572650
-0.77702868988581131
0.19425717247145291
0.19425717247145291
0.19425717247145291
0.19425717247145291
0.19425717247145291
0.19425717247145291

OpenBLAS 0.3.12 (macOS, MacPorts):
-0.38851434494290532
0.09712858623572633
0.09712858623572633
-0.77702868988581153
0.19425717247145272
0.19425717247145272
0.19425717247145272
0.19425717247145280
0.19425717247145280
0.19425717247145280

Apple (macOS 10.14.6):
-0.38851434494290549
0.09712858623572651
0.09712858623572648
-0.77702868988581120
0.19425717247145297
0.19425717247145297
0.19425717247145297
0.19425717247145297
0.19425717247145297
0.19425717247145297

MKL 2019.5 (Linux, Intel10_64lp, threaded code):
-0.38851434494290560
0.09712858623572651
0.09712858623572648
-0.77702868988581120
0.19425717247145294
0.19425717247145294
0.19425717247145294
0.19425717247145294
0.19425717247145294
0.19425717247145294
```

This was printed using the following code in the test:

```
    {
        int i;
        for (i=0; i < nodes; ++i) {
            printf("%.17f\n", MATRIX(V, i, 0));
        }
    }
```

I only see a failure with OpenBLAS. Notice that the last 6 elements of the eigenvector are identical with all LAPACK _except_ OpenBLAS, where the last 3 is very slightly different from the previous three. This is what is very likely causing the different result with OpenBLAS.

The problem is present only with the `IGRAPH_SCG_STOCHASTIC`/`IGRAPH_SCG_OPTIMUM` method, which is implemented in `igraph_i_optimal_partition` with the `matrix==3` choice.

**Comment by ntamas** [comment]
Fixed by #1568 so closing.
